### DISCUSSION TITLE
V2Ray的上传性能可能需要优化，使用透明代理直连，上传速度很慢

### DISCUSSION THREAD

**Discussion by lanchengjiu** [v2fly/v2ray-core#909] [Q&A]
我使用路由器搭建透明代理，路由器为小米R3D，CPU为IPQ8064，性能一般。
不经过V2Ray的情况下，使用www.speedtest.cn，选择最优节点进行测试，下载580M（ISP上限），上传60M
iptables配置REDIRECT转发流量到V2Ray，outbounds为直连，选择相同节点进行测试，下载180M，上传1.5M
下载速度还好一点，上传速度下降的太多了，我只能使用iptables+ipset代替V2Ray的routing
freedom直连的情况下，为什么经过V2Ray之后，上传速度会降低那么多，请问我应该如何优化？

另外，小米R3D是个极度精简的openwrt系统，系统本身没有libcap
我已经安装了libcap，在root的情况下运行V2Ray，sockopt的SO_MARK好像不生效，但是V2Ray的日志中没有提示失败。
V2Ray是否依赖libcap，或者其他的包？有劳了，谢谢各位大佬！

**Comment by jimsmt** [comment]
1.5Mbps可能是其他的问题吧，我iptables转发到透明代理，上传跑满宽带的上限200Mbps

**Reply by lanchengjiu** [reply]
你不会是软路由吧，路由CPU很好的话就没问题了。我用SOCKS5代理转发浏览器流量到PC本地V2Ray走直连出站，下载满速，上传80M，比路由器直连的43M还快，玄学了。上传为什么这么看路由器性能？但是实际上，上传的过程中路由器CPU并没有占用多少，不知道瓶颈在哪里。

**Comment by lanchengjiu** [comment]
测试方法：
使用测试PC访问www.speedtest.cn，选择最优节点进行测速，后续测试中不会更换测试节点。
小米路由器R3D配置透明代理，iptables表转发测试PC的流量到V2Ray，经过V2Ray内部从freedom直连出站

测试设备：
V2Ray 4.37.3、MiWiFi-R3D-2.29.5、IPQ8064、linux-arm32-v7a、Linux version 3.4.103

具体配置：
```
# iptables
iptables -t nat -N V2RAY
iptables -t nat -A V2RAY -s 192.168.1.100 -p tcp -j REDIRECT --to-ports 12345
iptables -t nat -A PREROUTING -p tcp -j V2RAY

# congif.json（最终简化到这个程度）
{
    "inbounds":[
        {
            "port":12345,
            "protocol":"dokodemo-door",
            "settings":{
                "network":"tcp",
                "followRedirect":true
            }
        }
    ],
    "outbounds":[
        {
            "protocol":"freedom"
        }
    ]
}

# 启动测试
killall -9 v2ray &> /dev/null
nohup /opt/v2ray/test/v2ray &
```

测试结果：
```
# 完全直连（ISP服务商的提供的宽带速度上限）
下载590 上传43

# iptables+ipset（白名单）=RETURN直连
下载596 上传43

# iptables+V2Ray=freedom直连（domainStrategy:AsIs）
额外损耗：inbounds:sniffing、routing:bittorrent+geosite:cn、outbounds:sockopt(mark)
下载131 上传1.6

# iptables+V2Ray=freedom直连（domainStrategy:AsIs）
额外损耗：inbounds:sniffing、routing:bittorrent+geosite:cn
下载135 上传1.6

# iptables+V2Ray=freedom直连（domainStrategy:AsIs）
额外损耗：inbounds:sniffing、routing:bittorrent
下载135 上传1.6

# iptables+V2Ray=freedom直连（domainStrategy:AsIs）
额外损耗：inbounds:sniffing
下载145 上传1.6

# iptables+V2Ray=freedom直连
下载134 上传1.6
```

总结结论：
无论怎么折腾V2Ray的配置，只要流量经过了V2Ray，就会大幅降速，尤其是上传速度。
是否开启嗅探、多种路由规则、出站增加mark，与最终速度关系不大。
在测试过程中，下载测速时路由器CPU满载，上传测速时CPU占用并不高，但是速度很慢。

由于一般境外VPS的线路速度都不会很快，所以代理情况下，对于V2Ray的性能要求其实不高。
但是直连情况下，国内ISP一般可以提供500M或1000M的下载速度，大约1/10的上传速度。
路由器上使用透明代理，由于路由CPU的性能比较低，再加上V2Ray的速度损耗，造成无法发挥国内ISP的宽带速度。

目前来看只能将直连分流放在V2Ray之外，通过iptables+ipset白名单进行路由分流，无法使用V2Ray内置的路由配置。
![微信图片_20210417162920](https://user-images.githubusercontent.com/43301062/115107539-0c016800-9f9e-11eb-9bef-ddd0291ed4a6.png)

**Comment by zw963** [comment]
楼主你好, 我个人一直在使用 ASUS 路由器做透明代理, 目前在用 RT-AC5300, 性能还算够用, 但是考虑到价格比较高, 一直考虑在路由器端寻找一个替代方案, 您可以针对(支持 V2Ray 作为客户端运行环境)的小米路由器给一个选择建议吗? 多谢.

**Reply by lanchengjiu** [reply]
没太懂你的需求，你说你在用RT-AC5300，用的也挺好。然后说想找个廉价方案替代RT-AC5300，让我推荐一个平价替换的设备吗？另外我看您是这个路由固件的第三方开发者，这么做是为了照顾帮助其他网友？

**Reply by zw963** [reply]
> 让我推荐一个平价替换的设备吗？

是的, 就是这个意思,  性能肯定越强越好, 价格呢, 比 ASUS 便宜一些, 最好有类似于 [Entwareentware](https://github.com/Entware/Entware) 的包管理, 但是如果预编译的 V2Ray 可以用, 也非必须吧. 



> 另外我看您是这个路由固件的第三方开发者，这么做是为了照顾帮助其他网友？

只是为了个人部署方便, 谈不上帮助了, 不过有需要的网友可以互相参考.

**Reply by lanchengjiu** [reply]
我玩的路由不多，没办法给推荐具体的路由，我只说说小米R3D上踩的坑吧。

- V2Ray编译好的包到不是问题，github上不就一大堆呢吗，不用自己编译，都是开箱即用
- R3D没有OpenWrt可用，所以做透明代理只能用Shell和Entware，但是Shell对于一般人门槛太高了。
- 就算用Shell，小米的精简系统内核也不支持TPROXY，既无法转发UDP，造成一些游戏的UDP透明代理失败，并且增加了DNS转发的难度（得去路由器的DNS服务器设置DNS请求转发）
- 估计是同上原因，SO_MARK有些问题，动OUTPUT会出现回环，造成路由器本机无法代理，不过这个倒是弃了也无大所谓
- 设备的CPU性能不大行，直连不能跑满500M上下行宽带，造成直连的路由分流得由外部iptables+ipset实现，又增加配置难度
- 综上所述，能刷OpenWrt，且CPU性能较好是比较重要的判定标准
